### PR TITLE
Fix problem with daily cron jobs triggered by force_cron_run

### PR DESCRIPTION
**Pull request by Zaoliang**
- see poo#31351 for more details
- verification run:
  http://e13.suse.de/tests/2518#step/force_cron_run/6

### CODE REVIEW

File: tests/console/force_cron_run.pm [hunk 1]
```diff
@@ -30,14 +30,17 @@ sub settle_load {
...
 
     # Make sure there's no load before we trigger one via cron.
     settle_load;
     my $before = time;
-    assert_script_run "bash -x /usr/lib/cron/run-crons", 1000;
+    #"bash -x /usr/lib/cron/run-crons" doesn't work really on a running system with online duration shorter than 15 minutes
+    # so we can call daily cron jobs explicitly here, no worries about other cron jobs for hourly,weekly,monthly because they
+    # are not available at that time
+    my $run_daily_cronjobs = 'for file in /etc/cron.daily/*; do echo execute daily cron jobs: $file; $file; done';
+    assert_script_run "$run_daily_cronjobs", 1000;
```
Comment (coolo) [inline tests/console/force_cron_run.pm]: No need for these quotes. It's just a variable.

But you need to be careful for the case when there are no cron.daily scripts installed. It would try to execute * - so either test -x $file before executing it in the loop. Or change the glob to e.g. for file in $(ls -1 /etc/cron.daily); /etc/cron.daily/$file; done

**Comment by okurz** [comment]
@coolo so is it really that we just want to run the daily crons? The btrfs jobs are weekly and monthly

**Comment by coolo** [comment]
No, actually we only care for those - and not about daily crons at all.

**Comment by coolo** [comment]
but I guess if we glob the .daily out as well, we are good to go :)

**Comment by okurz** [comment]
So rather cron.{hourly,daily,weekly,monthly}/* ?

**Comment by coolo** [comment]
should be enough I guess

**Comment by Zaoliang** [comment]
okay, updated, thanks for your input and comments.

**Comment by coolo** [comment]
http://e13.suse.de/tests/2521#step/force_cron_run/6 - so there are only daily ones?

**Comment by coolo** [comment]
CC @StefanBruens 

So let's drop this change in total - if Tumbleweed's support for cron.daily is buggy is basically uninteresting. It's obsolete anyway, we're migrating to systemd timers.

But we need to get these systemd timers triggered - and wait for their completion

File: tests/console/force_cron_run.pm [hunk 2]
```diff
@@ -30,13 +30,15 @@ sub settle_load {
...
     select_console 'root-console';
-
     # show dmesg output in console during cron run
     assert_script_run "dmesg -n 7";
 
     # Make sure there's no load before we trigger one via cron.
     settle_load;
     my $before = time;
+    # "bash -x /usr/lib/cron/run-crons" doesn't work really on a running system with online duration shorter than 15 minutes
+    # but what we can do is to disable all systemd timers. So keep bash -x /usr/lib/cron/run-crons for now
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: why should we disable the systemd timers? I thought the ideas was to explicitly trigger the timers which do basically the same as the run-crons command would do (on older systems). Second question then is if we should still call run-cron on top.

Comment (StefanBruens) [reply tests/console/force_cron_run.pm]: You can not trigger a timer (at least I am not aware of the possibility, even after studying the manpages), but only the service "behind" the trigger. I.e. if you have an hourly timer, it may still trigger again 5 minutes later.
This may be sufficient, e.g. it is likely running btrfs balance twice in a row is (mostly) a noop for the second run. There may be other triggers which cause the same amount of work each run.

Comment (okurz) [reply tests/console/force_cron_run.pm]: @Zaoliang @rwx788 I don't understand why do you want to disable services/timers?

@StefanBruens yes, I meant trigger services behind timers. And yes, I guess the idea is to trigger services explicitly and wait for their completion once. Any subsequent calls *should* be no-ops and the previous approach seems to have ensured this in principle.

Comment (coolo) [reply tests/console/force_cron_run.pm]: disabling might be the *only* way. As the trigger time is relative to the last time it ran, the monthly btrfs-balance.timer might hit in a saved HDD pretty much every test on a given day.

Comment (okurz) [reply tests/console/force_cron_run.pm]: But previously we also had the case that balancing could be triggered again but still a second run should not cause problems. Maybe we just need to improve how these problems show up in test results

Comment (coolo) [reply tests/console/force_cron_run.pm]: Then you are only talking about btrfs. So let's make this specific about this.

if rpm -q btrfsmaintenance ; then systemctl start btrfs-balance.service; fi
+ wait

**Comment by okurz** [comment]
ok, so let me try to summarize what we should do:

* For the SLE12 code stream instead of relying on the "runcron"-script we glob all cron jobs and run them explicitly
* For the SLE15+current code stream
  * We trigger all services for all timers instead
  * Of course we only trigger the existing services, e.g. if the package btrfsmaintenance is installed we trigger this one
  * We disable all timers to prevent them from retriggering at a random time during further test execution
* We will wait for the load to settle as we did in before

Correct?

File: tests/console/force_cron_run.pm (hunk 2 above)
Comment (rwx788) [reply tests/console/force_cron_run.pm]: I guess this is a must we should do, as btrfs balancing is the one hitting us most, I haven't seen issues cause by other services.

**Comment by rwx788** [comment]
@okurz this sounds to me.

File: tests/console/force_cron_run.pm (hunk 2 above)
Comment (okurz) [reply tests/console/force_cron_run.pm]:
```
$ rpm -ql btrfsmaintenance
/usr/lib/systemd/system/btrfs-balance.service
/usr/lib/systemd/system/btrfs-balance.timer
/usr/lib/systemd/system/btrfs-defrag.service
/usr/lib/systemd/system/btrfs-defrag.timer
/usr/lib/systemd/system/btrfs-scrub.service
/usr/lib/systemd/system/btrfs-scrub.timer
/usr/lib/systemd/system/btrfs-trim.service
/usr/lib/systemd/system/btrfs-trim.timer
/usr/lib/systemd/system/btrfsmaintenance-refresh.path
/usr/lib/systemd/system/btrfsmaintenance-refresh.service
…
```

File: tests/console/force_cron_run.pm [hunk 3]
```diff
@@ -30,14 +30,20 @@ sub settle_load {
...
     settle_load;
     my $before = time;
-    assert_script_run "bash -x /usr/lib/cron/run-crons", 1000;
+    # "bash -x /usr/lib/cron/run-crons" doesn't work really on a running system with online duration shorter than 15 minutes
+    # we stop and mask btrfs systemd timers. So keep bash -x /usr/lib/cron/run-crons for SLES 12
+	if (is_sle('<15') || is_leap('<15')) { 
+        assert_script_run("bash -x /usr/lib/cron/run-crons", 1000);
+    } 
+	else {
+		 assert_script_run('rpm -q btrfsmaintenance && systemctl start btrfs-balance.service && systemctl stop btrfs-* && systemctl mask btrfs-*');
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: why do you stop the service immediately after starting? I think you can only stop and mask the services after L49, i.e. after `settle_load`.

File: tests/console/force_cron_run.pm [hunk 4]
```diff
@@ -30,14 +30,20 @@ sub settle_load {
...
     assert_script_run "dmesg -n 7";
 
     # Make sure there's no load before we trigger one via cron.
     settle_load;
     my $before = time;
-    assert_script_run "bash -x /usr/lib/cron/run-crons", 1000;
+    # "bash -x /usr/lib/cron/run-crons" doesn't work really on a running system with online duration shorter than 15 minutes
+    # we stop and mask btrfs systemd timers. So keep bash -x /usr/lib/cron/run-crons for SLES 12
+	if (is_sle('<15') || is_leap('<15')) { 
+        assert_script_run("bash -x /usr/lib/cron/run-crons", 1000);
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: This does not solve the original problem as reported in the progress ticket though. Please see my recommendation in the comments

File: tests/console/force_cron_run.pm [hunk 5]
```diff
@@ -30,18 +30,27 @@ sub settle_load {
...
     my $before = time;
-    assert_script_run "bash -x /usr/lib/cron/run-crons", 1000;
+    # "bash -x /usr/lib/cron/run-crons" doesn't work really on a running system with uptime less than 15 minutes
+    # because no daily jobs got created or can be excecuted.
+    # we stop and mask btrfs systemd timers. So keep bash -x /usr/lib/cron/run-crons for SLES 12
+    if (is_sle('<15') || is_leap('<15')) {
+        assert_script_run("bash -x /usr/lib/cron/run-crons", 1000);
+    }
+    else {
+        assert_script_run('rpm -q btrfsmaintenance && systemctl start btrfs-balance.service');
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: I suggest to start *all* the timers to be more near to what we had for SLE12:

```
for i in $(sudo systemctl list-units --type=timer --state=active --no-legend -p=name | awk '{ print $1 }') ;do systemctl start $i ; done
```

there might be something nicer than awk but could not find a way to output only the name of the services easily. And we should not even need to check for the btrfsmaintenance package

@coolo this should ensure we do not need to hardcode service names and basically do the same as in before, run all timed jobs once. WDYT? Or should we rewrite the SLE12 case also to just point to specific services?

I don't think that btrfs-balance is enough because problems can be caused by the other btrfs maintenance jobs as well and also snapper(-cleanup), right?

File: tests/console/force_cron_run.pm [hunk 6]
```diff
@@ -30,18 +30,27 @@ sub settle_load {
...
-
     # show dmesg output in console during cron run
     assert_script_run "dmesg -n 7";
 
     # Make sure there's no load before we trigger one via cron.
     settle_load;
     my $before = time;
-    assert_script_run "bash -x /usr/lib/cron/run-crons", 1000;
+    # "bash -x /usr/lib/cron/run-crons" doesn't work really on a running system with uptime less than 15 minutes
+    # because no daily jobs got created or can be excecuted.
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: What exactly do you mean with "no daily jobs got created". Do you mean that symlinks in the folders /etc/cron.*/ are not yet there at the time of calling /usr/lib/cron/run-crons?

File: tests/console/force_cron_run.pm [hunk 7]
```diff
@@ -30,18 +30,27 @@ sub settle_load {
...
+    }
+    else {
+        assert_script_run('rpm -q btrfsmaintenance && systemctl start btrfs-balance.service');
+    }
     record_soft_failure 'bsc#1063638 - review I/O scheduling parameters of btrfsmaintenance' if (time - $before) > 60 && get_var('SOFTFAIL_BSC1063638');
     sleep 3;    # some head room for the load average to rise
     settle_load;
-
+    if !((is_sle('<15') || is_leap('<15'))) {
+        assert_script_run('rpm -q btrfsmaintenance && systemctl stop btrfs-* && systemctl mask btrfs-*');
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: Why should we stop the services? They should not even be running at this time. But what happens if you mask services that are currently running?

File: tests/console/force_cron_run.pm (hunk 5 above)
Comment (StefanBruens) [reply tests/console/force_cron_run.pm]: The 'sudo' is obviously not needed
'--property/-p' is only defined for the 'show' subcommand

File: tests/console/force_cron_run.pm [hunk 8]
```diff
@@ -15,7 +15,7 @@ use base "consoletest";
 use strict;
 use testapi;
 use utils 'assert_screen_with_soft_timeout';
-use version_utils 'is_jeos';
+use version_utils;
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: I guess you can back to the old line

File: tests/console/force_cron_run.pm [hunk 9]
```diff
@@ -30,18 +30,25 @@ sub settle_load {
 
 sub run {
     select_console 'root-console';
-
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: unrelated change, keep old

File: tests/console/force_cron_run.pm [hunk 10]
```diff
@@ -30,18 +30,25 @@ sub settle_load {
...
     # show dmesg output in console during cron run
     assert_script_run "dmesg -n 7";
 
     # Make sure there's no load before we trigger one via cron.
     settle_load;
     my $before = time;
-    assert_script_run "bash -x /usr/lib/cron/run-crons", 1000;
+    # run cron jobs or systemd timers which can affect system performance and mask systemd timers later
+    assert_script_run('find /etc/cron.{daily,weekly,monthly} -type f -executable -exec echo cron job: {} \; -exec {} \;');
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: why not hourly?

File: tests/console/force_cron_run.pm [hunk 11]
```diff
@@ -30,18 +30,25 @@ sub settle_load {
...
+    assert_script_run(
+        'for i in $(systemctl list-units --type=timer --state=active --no-legend | awk \'{ print $1 }\' | sed \'s/.timer//\'); do
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: Because we use sed anyway we can simplify this line to

```perl
for i in $(systemctl list-units --type=timer --state=active --no-legend | sed \'s/.timer.*$//\'); do
```

File: tests/console/force_cron_run.pm [hunk 12]
```diff
@@ -30,18 +30,25 @@ sub settle_load {
...
+    echo "Triggering systemd timed service $i" && systemctl start $i ; done'
```
Comment (okurz) [inline tests/console/force_cron_run.pm]: Just immediately mask the service here as well so then we do not even have any duplication further down

File: tests/console/force_cron_run.pm (hunk 11 above)
Comment (Zaoliang) [reply tests/console/force_cron_run.pm]: awk \'{ print $1 }\' is required, otherwise it will fail:

http://e13.suse.de/tests/6094#step/force_cron_run/11

File: tests/console/force_cron_run.pm [hunk 13]
```diff
@@ -30,18 +30,21 @@ sub settle_load {
...
+    # run cron jobs or systemd timers which can affect system performance and mask systemd timers later
+    assert_script_run('find /etc/cron.{hourly,daily,weekly,monthly} -type f -executable -exec echo cron job: {} \; -exec {} \;');
+    assert_script_run(
+        'for i in $(systemctl list-units --type=timer --state=active --no-legend | awk \'{ print $1 }\' | sed \'s/.timer//\'); do
+    echo "Triggering systemd timed service $i" && systemctl start $i && systemctl mask $i ; done'
+    );
     record_soft_failure 'bsc#1063638 - review I/O scheduling parameters of btrfsmaintenance' if (time - $before) > 60 && get_var('SOFTFAIL_BSC1063638');
     sleep 3;    # some head room for the load average to rise
     settle_load;
-
```
Comment (SergioAtSUSE) [inline tests/console/force_cron_run.pm]: I would apply the same here. Keep old ;)

File: tests/console/force_cron_run.pm (hunk 11 above)
Comment (StefanBruens) [reply tests/console/force_cron_run.pm]: Unfortunately I can not see the error from outside SUSE ...

You may try `sed -e 's/\(\S\+\)\.timer\s.*/\1/'`

Comment (Zaoliang) [reply tests/console/force_cron_run.pm]: yes, I checked your suggestion, it is working, thanks!

**Comment by okurz** [comment]
Please crosscheck carefully if we follow the suggestions in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/5097#issuecomment-400621099 as well. I still think that we should not mask the services in the "system_performance" scenarios

**Comment by SergioAtSUSE** [comment]
As spoken with @okurz, merging this PR and working on other PR to handle system_performance scenarios:
- https://openqa.opensuse.org/tests?match=system_performance
- and renaming of the module (it is no more "force cron run" only
